File: 1312-minimum-insertion-steps-to-make-a-string-palindrome/1312-minimum-insertion-steps-to-make-a-string-palindrome.py
class Solution(object):
    def minInsertions(self, s):
        """
        :type s: str
        :rtype: int
        """
        if s == s[::-1] : return 0
        n = len(s)
        dp = [[0]*n for _ in range(n)]

        for left in range(n-1, -1, -1) :
            dp[left][left] = 1
            for right in range(left+1,len(s)) : 
                if s[left] == s[right] :
                    dp[left][right] = 2 + dp[left+1][right-1]
                else :
                    dp[left][right] = max(dp[left][right-1], dp[left+1][right])

        return len(s) - dp[0][len(s)-1]

File: 1312-minimum-insertion-steps-to-make-a-string-palindrome/1312-minimum-insertion-steps-to-make-a-string-palindrome.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// int strLength(char* s) {
//     int length = 0;
//     while (s[length] != '\0') length++;
//     return length;
// }

int minInsertions(char* s) {
    int n = strlen(s);
    int **dp = (int**)malloc(n * sizeof(int*));
    for (int i = 0; i < n; i++) {
        dp[i] = (int*)malloc(n * sizeof(int));
        memset(dp[i], 0, n * sizeof(int)); 
    }

    for (int left = n - 1; left >= 0; left--) {
        dp[left][left] = 1;
        for (int right = left + 1; right < n; right++) {
            if (s[left] == s[right]) {
                dp[left][right] = 2 + dp[left + 1][right - 1];
            } else {
                dp[left][right] = (dp[left][right - 1] > dp[left + 1][right]) ? dp[left][right - 1] : dp[left + 1][right];
            }
        }
    }

    int result = n - dp[0][n - 1];

    for (int i = 0; i < n; i++) {
        free(dp[i]);
    }
    free(dp);

    return result;
}